sdk/python/feast/go_server.py

```diff
@@ -0,0 +1,246 @@
+        atexit.register(lambda: self.stop())
+        self.start_grpc_server()
+        self.pipe_closed = False
+
+    def start_grpc_server(self):
+        """Start the gRPC server on the Go subprocess."""
+        if self.grpc_server_started:
+            return
+
+        # Try 5 times to wait for pipe to open
+        for i in range(5):
+            try:
+                self.process.stdin.write(b"startGrpc\n")
+                self.process.stdin.flush()
+                self.grpc_server_started = True
+                break
+            except subprocess.CalledProcessError as error:
+                # If there is an exception the Go subprocess is probably closed
+                # so the pipe is also probably closed.
+                self.pipe_closed = True
+                self.stop()
+                raise GoSubprocessConnectionFailed() from error
+
+        channel = grpc.insecure_channel(f"127.0.0.1:{self.unused_port}")
+        self.client = ServingServiceStub(channel)
+
+        # Make sure the connection can be used for feature retrieval before returning
+        # from constructor. We try connecting to the Go subprocess for 5 seconds (50
+        # times with 0.1 second intervals).
+        # TODO (Ly): Review: increase time.sleep(0.2) to 0.2 so wait 10 seconds
+        # since python connector plugin can take up to 8 seconds to spin up.
+        # 5 seconds is not enough sometimes to spin up grpc server itself
+        for i in range(50):
+            try:
+                self.client.GetFeastServingInfo(request=GetFeastServingInfoRequest())
+                break
+            except Exception as e:
+                if i == 49:
+                    self.stop()
+                    raise GoSubprocessConnectionFailed() from e
+
+            # Sleep for 0.1 second before retrying.
+            time.sleep(0.1)
+
+    def stop(self):
+        """Stop the gRPC server on the Go subprocess and ensure the process is dead."""
+        if not self.pipe_closed:
+            try:
+                # Let the Go subprocess clean up and shut down itself.
+                self.process.stdin.write(bytes("stop\n", encoding="utf8"))
+                self.process.stdin.flush()
+                # TODO (Ly): Review: We don't close stdin here
+                # since if the call succeeds go process closes
+                # itself and stdin?
+                # self.process.stdin.close()
+            except subprocess.CalledProcessError as error:
+                # If there is a problem telling the Go subprocess to stop,
+                # directly terminate it.
+                if not isinstance(error, BrokenPipeError):
+                    self.process.terminate()
+                else:
+                    raise
+
+        self.grpc_server_started = False
+        self.pipe_closed = True
+
+
+def _get_unused_port() -> str:
+    sock = socket.socket()
+
+    # Binding to port 0 means the OS will choose an unused port
+    sock.bind(("", 0))
```

Comment:
We probably don't want to just randomly select a port. Otherwise we are adding non-determinism into the our users' environments. I think it would be more sane to start at a specific port value (something high) and fail if its used (which would indicate that multiple versions of this server are running). Or is there a use case for multiple Go servers?

Reply:
If there are multiple python processes running feast (in different shell sessions) all use go_server, then we need multiple go_server subprocesses. We could start something high, i.e: 54321, and add 1 if it's used until we find an unused port?

Reply:
sgtm